mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/user/PostgresConnectionResolver.java

```diff
@@ -83,4 +98,53 @@ public Mono<PostgresqlConnection> resolver(Domain domain) {
         }
     }
 
+    class PostgresConnectionPoolResolver implements PostgresConnectionResolver, Closeable {
+        private static final Logger LOGGER = LoggerFactory.getLogger(PostgresConnectionPoolResolver.class);
+
+        private final ConnectionPool connectionPool;
+
+        private final Map<String, Connection> mapConnectionLookup;
+
+        public PostgresConnectionPoolResolver(ConnectionPoolConfiguration poolConfiguration) {
+            this.connectionPool = new ConnectionPool(poolConfiguration);
+
+            this.mapConnectionLookup = new HashMap<>();
+        }
+
+        @Override
+        public Mono<? extends Connection> resolver(Domain domain) {
+            return Mono.fromCallable(() -> mapConnectionLookup.get(domain.asString()))
+                .switchIfEmpty(createPooledConnection(domain)
+                    .doOnSuccess(connection -> mapConnectionLookup.put(domain.asString(), connection)));
+        }
+
+        @Override
+        public Mono<Void> release(Domain domain) {
+            return Mono.fromCallable(() -> mapConnectionLookup.get(domain.asString()))
+                .flatMap(release());
+        }
+
+        private Mono<Connection> createPooledConnection(Domain domain) {
+            return connectionPool.create()
+                .doOnNext(e -> System.out.println("connection has been created, " + e.hashCode()))
+                .flatMap(pooledConnection ->
+                    Mono.from(pooledConnection.createStatement("SET app.current_domain TO '" + domain.asString() + "'")
+                            .execute())
+                        .thenReturn(pooledConnection));
+        }
+
+        private static Function<Connection, Mono<Void>> release() {
+            return connection -> Mono.from(connection.createStatement("RESET app.current_domain").execute())
+                .doOnSuccess(e -> {
+                    LOGGER.debug("connection has been released, {}", e.hashCode());
+                })
+                .then(Mono.from(connection.close()));
```

Comment:
Hmm keep disposing and creating connection again for every business query seems not good to me.

Reply:
`connection` here is `pooledConnection`, one wrapper, 
`.close` method in `pooledConnection` is "release connection to the pool", it does not `close` the physical connection. 

//I same confused when first use it